libraries/tools/kotlin-gradle-plugin/src/main/kotlin/org/jetbrains/kotlin/gradle/tasks/Tasks.kt

```diff
@@ -168,14 +170,14 @@ abstract class AbstractKotlinCompile<T : CommonCompilerArguments>() : AbstractKo
         get() = (classpath + additionalClasspath)
             .filterTo(LinkedHashSet(), File::exists)
 
-    private val sourceFilesExtensionsSources: MutableList<Iterable<String>> = mutableListOf()
+    private val sourceFilesExtensionsSources = project.objects.listProperty(String::class.java)
 
     @get:Input
     val sourceFilesExtensions: List<String>
-        get() = DEFAULT_KOTLIN_SOURCE_FILES_EXTENSIONS + sourceFilesExtensionsSources.flatten()
+        get() = DEFAULT_KOTLIN_SOURCE_FILES_EXTENSIONS + sourceFilesExtensionsSources.get()
 
     internal fun sourceFilesExtensions(extensions: Iterable<String>) {
-        sourceFilesExtensionsSources.add(extensions)
+        sourceFilesExtensionsSources.addAll(extensions)
```

Comment:
Another problem is that `ListProperty#addAll` was added in Gradle 4.10. We cannot drop support for all versions below 4.10.